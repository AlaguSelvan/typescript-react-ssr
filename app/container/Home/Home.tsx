import React, { useState, useEffect } from 'react';
import axios from 'axios';
import { useSelector, connect } from 'react-redux';
import { Link } from 'react-router-dom';
import Button from '../../components/Button';
import * as HomeActions from '../../redux/home/actions';

const Home = ({ userData, fetchInitialDataIfNeeded }: any) => {
  const [data, setData] = useState([]);
  const [isFetching, setIsFetching] = useState(false);

  const fetchUserData = async () => {
    setIsFetching(true);
    const { data } = await axios.get(
      'https://jsonplaceholder.typicode.com/users'
    );
    setData(data);
    setIsFetching(false);
  };

  useEffect(() => {
    if (userData.length === 0) {
      fetchInitialDataIfNeeded();
    }
  }, []);

  return (
    <div>
      <p>Home Page...</p>
      <div style={{ display: 'flex' }}>
        <div>
          <b>Initial Preloaded Data</b>
          <ul>
            {userData.map(({ name }: any, index: number) => (
              <li key={index}>{name}</li>
            ))}
          </ul>
        </div>
        <div>
          <div style={{ display: 'flex', flexDirection: 'column' }}>
            <b>Data from Client</b>
            {!isFetching ? (
              <button onClick={() => fetchUserData()}>Fetch Data</button>
            ) : (
              <p>Fetching</p>
            )}
          </div>
          {!isFetching ? (
            <ul>
              {data.map(({ name }: any, index: number) => (
                <li key={index}>{name}</li>
              ))}
            </ul>
          ) : (
            <p>No Data</p>
          )}
        </div>
        {/* <Button /> */}
      </div>
      <Link to="/about">
        <p>goto next ðŸ‘‰ page</p>
      </Link>
    </div>
  );
};

const mapStateToProps = ({ HomeReducer: { userData } }: any) => ({
  userData
});

const mapDispatchToProps = (dispatch: any) => ({
  fetchInitialDataIfNeeded: () => dispatch(HomeActions.fetchUserData())
});

export default connect(mapStateToProps, mapDispatchToProps)(Home);
