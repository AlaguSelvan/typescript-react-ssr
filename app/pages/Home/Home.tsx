import React, { Component } from 'react'
import { Link } from 'react-router-dom'

const Home = () => (
  <div>
    <p>
    Hello World...
    </p>
    <Link to="/about">
    <p>goto Next page</p>
      </Link>
  </div>
)


export default Home