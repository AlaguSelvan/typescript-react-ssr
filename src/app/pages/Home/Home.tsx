import React, { Component } from 'react'

export const Home = (
  <div>
    Hello World...
  </div>
)
