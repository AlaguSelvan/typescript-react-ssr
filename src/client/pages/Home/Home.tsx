import React, { Component } from 'react'

const Home = () => (
  <div>
    Hello World...
  </div>
)


export default Home