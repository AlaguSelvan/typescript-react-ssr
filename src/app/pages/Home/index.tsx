import * as React from 'react'
import {Home} from './Home';

export default Home;
