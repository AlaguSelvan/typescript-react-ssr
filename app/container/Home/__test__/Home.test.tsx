import React from 'react';
import { render, cleanup, waitForElement } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import { Home } from '../Home';
import IUserData from '../../../types/UserData';

afterEach(cleanup);

const fakeData = [
  {
    id: 1,
    name: 'Leanne Graham',
    username: 'Bret',
    email: 'Sincere@april.biz',
    address: {
      street: 'Kulas Light',
      suite: 'Apt. 556',
      city: 'Gwenborough',
      zipcode: '92998-3874',
      geo: {
        lat: '-37.3159',
        lng: '81.1496'
      }
    },
    phone: '1-770-736-8031 x56442',
    website: 'hildegard.org',
    company: {
      name: 'Romaguera-Crona',
      catchPhrase: 'Multi-layered client-server neural-net',
      bs: 'harness real-time e-markets'
    }
  }
];

interface IProps {
  userData: IUserData[];
}

interface IActions {
  fetchInitialDataIfNeeded: () => null;
}

describe('<Home />', () => {
  const tree = (props: IProps, actions: IActions) => {
    render(
      <MemoryRouter>
        <Home {...props} {...actions} />
      </MemoryRouter>
    ).container.firstChild;
  };

  it('should call fetchInitialDataIfNeeded when loaded', () => {
    const mockAction = jest.fn();
    const actions = { fetchInitialDataIfNeeded: mockAction };
    const userData = [];
    tree({ userData }, actions);
    expect(mockAction).toHaveBeenCalled();
  });

  it('should not call fetchInitialDataIfNeeded when loaded on Server Side', () => {
    const mockAction = jest.fn();
    const actions = { fetchInitialDataIfNeeded: mockAction };
    const userData = fakeData;
    tree({ userData }, actions);
    expect(mockAction).not.toHaveBeenCalled();
  });

  it('renders the loading status if requesting data', () => {
    const userData = fakeData;
    const actions = { fetchInitialDataIfNeeded: () => null };

    expect(tree({ userData }, actions)).toMatchSnapshot();
  });

  it('renders the list if loading was successful', () => {
    const userData = fakeData;
    const actions = { fetchInitialDataIfNeeded: () => null };

    expect(tree({ userData }, actions)).toMatchSnapshot();
  });
});
