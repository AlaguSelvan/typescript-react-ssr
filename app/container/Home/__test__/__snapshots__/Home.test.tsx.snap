// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders the <UserList /> if loading was successful 1`] = `undefined`;

exports[`<Home /> renders the list if loading was successful 1`] = `undefined`;

exports[`<Home /> renders the loading status if requesting data 1`] = `undefined`;
